pingcap/tidb
# tikv: support cancel before cop response return

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/18057

continue to https://github.com/pingcap/tidb/pull/18277

Problem Summary:

now cop task maybe cannot exit due to goroutine blocked in grpc and could not see kill flag

### What is changed and how it works?

What's Changed, How it Works:

we cannot cancel root context, but we can cancel "context used to send rpc"

so this PR chooses to collect CancelFunc for each RPCs in a cop query and cancel them during executor closing.

### Related changes

- n/a

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

```
1. start a cluster with pd and tikv(need inject in batchclient which need tikv instead of mocktikv or unistore)
2. start two mysql sessions: A and B
3. `curl -XPUT -d 'return(true)' http://0.0.0.0:10080/fail/github.com/pingcap/tidb/store/tikv/mockBlockOnBatchClient` to enable failpoint which cause every kv request will block 1 hour
4. execute a cop SQL like `select * from t` it will blocked in session A
5. `show processlist` and `kill tidb query N` in session B
```

- without 40b7b26 will see query cannot be kill and goroutine blocked in ticlient and recvFromRespCh
- without other's will see goroutine block in ticlient and copIterator.Close
- with this pr will see query can be killed and report error

Side effects

- n/a

### Release note 

- support cancel before RPC response return.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/18445)
